R/importM.COSMIC.R

```diff
@@ -0,0 +1,348 @@
+    }
+
+    # Read header line from fNameCNV file
+    tmpHeader <- readLines(fNameCNV, n = 1)
+
+    # Create character vector of col names from header string
+    header <- unlist(strsplit(tmpHeader[1], "\t"))
+
+    # The required columns from Cosmic CNV table
+    required <- c("ID_GENE", "ID_SAMPLE", "ID_TUMOUR", "TOTAL_CN")
+
+    # Find the column numbers in the header that match headings in required
+    iCol <- which(header %in% required)
+
+    # Create a column mask for readr::read_delim()
+    readMask <- paste(paste(rep("_", length(header)),
+                            collapse = ""),
+                      sep = "")
+
+    # Set target columns to characters
+    for (col in iCol) {
+        substr(readMask, col, col) <- "c"
+    }
+
+    # Read only the selected columns from Cosmic CNV file
+    dataCNV <- readr::read_delim(file = fNameCNV,
+                                 delim = "\t",
+                                 col_types = readMask,
+                                 n_max = Inf)
+
+    if (!silent) {
+        cat("Done reading from ", fNameCNV, ".\n")
+    }
+
+    # ==== MAP COSMIC IDs TO HGNC GENE SYMBOLS ================================
+
+    ## Used the following call to generate fastMapCosmic.rds:
+    ## fastMapGenerate("hgnc_complete_set.txt", "symbol",
+    ##                 "cosmic", type = "Cosmic",
+    ##                 outputName = "../inst/extdata/fastMapCosmic.rds")
+
+    # load fastMap hash table
+    fastMapCosmic <- readRDS(system.file("extdata",
+                                       "fastMapCosmic.rds",
+                                       package = "rete"))
+
+    # Validate that the type of the hash table is "Cosmic"
+    if (!is.null(attributes(fastMapCosmic)$type)) {
+        if ("Cosmic" != attributes(fastMapCosmic)$type) {
+            stop(sprintf("Expected hash table type: Cosmic.\nSupplied hash table type: %s",
+                         attributes(fastMapCosmic)$type))
+        }
+    } else {
+        stop("Supplied hash table does not have a type attribute.")
+    }
+
+
+    # ==== CONSTRUCT rCNA OBJECT ==============================================
+
+    if (!silent) {
+        cat("Generating rCNA object using CNA data in ", fNameCNV, ".\n")
+    }
+
+    barcodes <- c()
+    hgncSym <- c()
+
+    for (i in 1:nrow(dataCNV)) {
+
+        # Generate all barcodes for the mutations in dataCNV
+        barcodes <- append(barcodes,
+                           paste("TCGA-",
+                                 dataCNV[i,]$ID_SAMPLE, "-",
+                                 dataCNV[i,]$ID_TUMOUR,
+                                 sep = ""))
+
+        # Generate all HGNC symbols using fastMap()
+        hgncSym <- append(hgncSym,
+                          fastMap(dataCNV[i,]$ID_GENE,
+                                  fastMapCosmic,
+                                  type = "Cosmic",
+                                  dev = TRUE))
+
+    }
```

Comment:
And at the end of this loop you have a vector of 29,064,853 gene symbols and 29,064,853 barcodes...

This whole loop must go. All of this is trivially vectorized. (Code below, but untested)
```
hgncSym <- unique(dataCNV$gene_name)
dataCNV$barcode <- paste("TCGA", dataCNV$ID_SAMPLE, dataCNV$ID_TUMOUR, sep = "-" )
barcodes <- unique(dataCNV$barcode)
```
This also creates a new column, which you'll need later.